tokio/src/time/driver/entry.rs

```diff
@@ -1,361 +1,747 @@
+    pub(crate) fn new(handle: &InternalHandle<TS>, deadline: Instant) -> TimerEntry<TS> {
+        let deadline = handle.time_source().deadline_to_tick(deadline);
+        let driver = handle.clone();
 
-        if is_elapsed(curr) {
-            return Poll::Ready(if curr == ERROR {
-                Err(Error::from_u8(self.error.load(SeqCst)))
-            } else {
-                Ok(())
-            });
+        Self {
+            driver,
+            inner: StdUnsafeCell::new(TimerShared {
+                waker: AtomicWaker::new(),
+                state: state_cell::StateCell::new(),
+                padded: TimerSharedPadded::new(deadline),
+                _p: PhantomPinned,
+            }),
         }
+    }
+
+    fn inner(&self) -> &TimerShared {
+        unsafe { &*self.inner.get() }
+    }
 
-        Poll::Pending
+    /// Gets an EntryHandle to this Entry
+    pub(super) fn handle(&mut self) -> TimerHandle {
+        TimerHandle {
+            inner: NonNull::new(self.inner.get()).unwrap(),
+        }
+    }
+
+    pub(crate) fn is_elapsed(&self) -> bool {
+        self.inner().state.get(Ordering::Relaxed).is_elapsed()
     }
 
-    /// Only called by `Registration`
-    pub(crate) fn reset(entry: &mut Arc<Entry>) {
-        let inner = match entry.upgrade_inner() {
-            Some(inner) => inner,
-            None => return,
+    /// Cancels and deregisters the timer. This operation is irreversible.
+    pub(crate) fn cancel(self: Pin<&mut Self>) {
+        // We need to perform an acq/rel fence with the driver thread, and the
+        // simplest way to do so is to grab the driver lock.
+        //
+        // Why is this necessary? We're about to release this timer's memory for
+        // some other non-timer use. However, we've been doing a bunch of
+        // relaxed (or even non-atomic) writes from the driver thread, and we'll
+        // be doing more from _this thread_ (as this memory is interpreted as
+        // something else).
+        //
+        // It is critical to ensure that, from the point of view of the driver,
+        // those future non-timer writes happen-after the timer is fully fired,
+        // and from the purpose of this thread, the driver's writes all
+        // happen-before we drop the timer. This in turn requires us to perform
+        // an acquire-release barrier in _both_ directions between the driver
+        // and dropping thread.
+        //
+        // This lock acquisition serves this purpose. All of the driver
+        // manipulations happen with the lock held, so we can just take the lock
+        // and be sure that this drop happens-after everything the driver did so
+        // far and happens-before everything the driver does in the future.
+        // While we have the lock held, we also go ahead and deregister the
+        // entry if necessary.
+        unsafe {
+            self.driver
+                .clear_entry(NonNull::new(self.inner() as *const _ as *mut _).unwrap())
```

Comment:
nit: I believe there's a `From<&'_ T> for NonNull<T>`?
```suggestion
                .clear_entry(NonNull::from(self.inner()).unwrap())
```